feat: create ages system

GPerino opened the pull request.

GPerino committed:
feat : Add card list in json
- Add structure per age
#24

GPerino committed:
feat : Add age system
- First create of age & card
- Add max turn constant in rule file
- Add function to load json
- Add basic game loop for age system
#24

GPerino committed:
feat : Add age system
- Remove for now json files
- Remove useless constants in rules ( & tests use it)
- Remove useless functions in Age (now  just contains constructor)
- Remove for now Card object
#24

GPerino committed:
fix : lint problem from last commit
#24

GPerino committed:
Merge branch 'develop' into 24-create-ages-system

# Conflicts:
#	src/domain/engine/Master/master.ts
#	src/domain/engine/Rule/rule.test.ts
#	src/domain/engine/Rules/rules.ts
#	tsconfig.json

GPerino committed:
refactor : use dependencies injection in game loop object & tests
- Remove Rule file ( replaced by Rules)
#24

vbetsch commented on src/domain/engine/Game/age.ts:
```diff
@@ -0,0 +1,12 @@
+export class Age {
+  public _structure: Array<boolean> = [];
+  public _cards: Array<string> = [];
```
You need to have private properties and getters to get private properties.
And the convention is : starting with underscore for private properties.
We use `type[]` instead of `Array<type>` but both are accepted.

vbetsch commented on src/domain/engine/Game/age.ts:
```diff
@@ -0,0 +1,12 @@
+export class Age {
+  public _structure: Array<boolean> = [];
```
structure and cards are not required for this development but if it's working it's accepted

vbetsch commented on src/domain/engine/Game/game-loop.test.ts:
```diff
@@ -0,0 +1,35 @@
+    container.clearInstances();
+  });
+  it('should run the expected number of ages', () => {
+    const loop: GameLoop = new GameLoop();
```
loop has to be defined in `beforeEach` in order to avoid border effects

vbetsch commented on src/domain/engine/Game/game-loop.test.ts:
```diff
@@ -0,0 +1,35 @@
+  let rules: Rules;
+  let mockLogger: jest.Mocked<Logger>;
+  beforeEach(() => {
+    rules = container.resolve(Rules);
```
Rules need to be mocked

vbetsch commented on src/domain/engine/GameLoop/game-loop.ts:
You need to create a folder for GameLoop.
The convention is : one class by folder.

vbetsch commented on src/domain/engine/Game/game-loop.ts:
```diff
@@ -0,0 +1,21 @@
+  }
+
+  public start(): void {
+    for (let i: number = 1; i < this._rules.agesNumber + 1; i++) {
```
We prefer `<= this._rules.agesNumber` instead of `< this._rules.agesNumber + 1` (like in its test)

vbetsch commented on src/domain/engine/Game/game-loop.ts:
```diff
@@ -0,0 +1,21 @@
+  public start(): void {
+    for (let i: number = 1; i < this._rules.agesNumber + 1; i++) {
+      this._logger.log('age ' + i + '  started');
+      //implement turn by turn functionalities
```
please avoid comments

vbetsch commented on src/domain/engine/Game/game-loop.ts:
```diff
@@ -0,0 +1,21 @@
+    for (let i: number = 1; i < this._rules.agesNumber + 1; i++) {
+      this._logger.log('age ' + i + '  started');
+      //implement turn by turn functionalities
+      this._logger.log('age ' + i + '  finished');
```
The string is not exactly the same but it's accepted

vbetsch commented on src/domain/engine/Game/age.ts:
You need to create a class-folder and add a test on it ;)

vbetsch commented on src/domain/engine/Game/game.test.ts:
```diff
@@ -1,3 +1,4 @@
+import 'reflect-metadata';
```
You don't need this import. You have to delete it in order to optimize performances.

vbetsch reviewed (changes requested):
Now the conception is ok, but the good pratices are not followed

vbetsch commented:
I just checked one thing : Age is not used. So there is a bug in conception.

GPerino committed:
refactor: some changes from pull request comments
- Remove unused age class
- Mock rule object in test
- create folder foreach classes
#24

vbetsch commented on src/domain/engine/GameLoop/game-loop.test.ts:
```diff
@@ -4,32 +4,37 @@ import { Logger } from '@core/Logger/logger';
     container.registerInstance(Logger, mockLogger);
+
+    mockRules = {
+      agesNumber: 3, // valeur de test
```
Please no comments in French ! I'm going to have to ban them now :(

vbetsch reviewed (changes requested):
Please don't put comments when it's not necessary, especially not in French !

GPerino committed:
refactor: remove unnecessary comment
#24

vbetsch reviewed (approved):
Nice !
